datadog_checks_dev/datadog_checks/dev/tooling/manifest_validator/validator.py

```diff
@@ -0,0 +1,523 @@
+        }
+    )
+
+
+def is_metric_in_metadata_file(metric, check):
+    """
+    Return True if `metric` is listed in the check's `metadata.csv` file, False otherwise.
+    """
+    metadata_file = get_metadata_file(check)
+    if not os.path.isfile(metadata_file):
+        return False
+    for _, row in read_metadata_rows(metadata_file):
+        if row['metric_name'] == metric:
+            return True
+    return False
+
+
+ValidationResult = namedtuple('ValidationResult', 'failed fixed messages')
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ManifestValidator(object):
+    def __init__(self, is_extras=False, is_marketplace=False):
+        self.result = ValidationResult(
+            failed=False, fixed=False, messages={'success': [], 'warning': [], 'failure': [], 'info': []}
+        )
+        self.is_extras = is_extras
+        self.is_marketplace = is_marketplace
+
+    def validate(self, check_name, manifest, should_fix):
+        # type: (str, Dict, bool) -> ValidationResult
+        """Validates the decoded manifest. Will perform inline changes if fix is true"""
+        raise NotImplementedError
+
+    def fail(self, error_message):
+        self.result.failed = True
+        self.result.messages['failure'].append(error_message)
+
+    def fix(self, problem, solution):
+        self.result.warning_msg = problem
+        self.result.success_msg = solution
+        self.result.fixed = True
+        self.result.failed = False
+
+
+class AttributesValidator(ManifestValidator):
+    """ attributes are valid"""
+
+    def validate(self, check_name, decoded, fix):
+        errors = sorted(get_manifest_schema().iter_errors(decoded), key=lambda e: e.path)
+        if errors:
+            for error in errors:
+                self.fail(f'  {"->".join(map(str, error.absolute_path))} Error: {error.message}')
+
+
+class GUIDValidator(ManifestValidator):
+    all_guids = {}
+
+    def validate(self, check_name, decoded, fix):
+        guid = decoded.get('guid')
+        if guid in self.all_guids:
+            output = f'  duplicate `guid`: `{guid}` from `{self.all_guids[guid]}`'
+            if fix:
+                new_guid = uuid.uuid4()
+                self.all_guids[new_guid] = check_name
+                decoded['guid'] = new_guid
+                self.fix(output, f'  new `guid`: {new_guid}')
+            else:
+                self.fail(output)
+        elif not guid or not isinstance(guid, str):
+
+            output = '  required non-null string: guid'
+            if fix:
+                new_guid = uuid.uuid4()
+                self.all_guids[new_guid] = check_name
+                decoded['guid'] = new_guid
+                self.fix(output, f'  new `guid`: {new_guid}')
+            else:
+                self.fail(output)
+        else:
+            self.result.failed = True
+            self.all_guids[guid] = check_name
+        return self.result
+
+
+class ManifestVersionValidator(ManifestValidator):
+    root = get_root()
```

Comment:
this won't be set at load time

Reply:
It is when invoked (as in validation passes) but not on the test (and that's why is broken)

Reply:
The test passes locally, do you know why root is `foo` on ci?

```
E   AssertionError: 
E     
E       metrics_metadata in manifest.json references a non-existing file: /foo/active_directory/metadata.csv.
E     
E   assert not True
E    +  where True = \n\n  metrics_metadata in manifest.json references a non-existing file: /foo/active_directory/metadata.csv.\n.failed
E    +    where \n\n  metrics_metadata in manifest.json references a non-existing file: /foo/active_directory/metadata.csv.\n = \n\n  metrics_metadata in manifest.json references a non-existing file: /foo/active_directory/metadata.csv.\n.result
```

https://dev.azure.com/datadoghq/integrations-core/_build/results?buildId=57611&view=logs&j=910f82a4-622d-5ace-f877-8b75c045486c&t=dcbfe806-1c22-5885-41c0-81f1272e0603&l=433

Reply:
what do you mean?

Reply:
If you run `dded validate manifest` with this code validation passes.

Reply:
But it should be moved to `__init__`

Reply:
ah it was working b/c ci enters the root already